packages/core/src/caches/index.ts

```diff
@@ -56,6 +45,107 @@ export class RedisCache implements CacheStore {
       consoleLog.info('[CACHE] Disconnected from Redis');
     }
   }
+
+  protected getSocketOptions(url: URL) {
+    const certFile = url.searchParams.get('cert');
+    const keyFile = url.searchParams.get('key');
+    const caFile = url.searchParams.get('certroot');
+
+    return {
+      rejectUnauthorized: yes(url.searchParams.get('reject_unauthorized')),
+      tls: url.protocol === 'rediss',
+      cert: certFile ? fs.readFileSync(certFile).toString() : undefined,
+      key: keyFile ? fs.readFileSync(keyFile).toString() : undefined,
+      ca: caFile ? fs.readFileSync(caFile).toString() : undefined,
+      reconnectStrategy: (retries: number, cause: Error) => {
+        if ('code' in cause && cause.code === 'SELF_SIGNED_CERT_IN_CHAIN') {
+          // This will throw only if reject unauthorized is true (default).
+          // Doesn't make sense to retry.
+          return false;
+        }
+
+        return Math.min(retries * 50, 500);
+      },
+    };
+  }
+
+  protected abstract ping(): Promise<string | undefined>;
+}
+
+export class RedisCache extends RedisCacheBase {
+  readonly client?: RedisClientType;
+
+  constructor(redisUrl?: string | undefined) {
+    super();
+
+    if (redisUrl) {
+      this.client = createClient({
+        url: conditional(!yes(redisUrl) && redisUrl),
+        socket: trySafe(() => this.getSocketOptions(new URL(redisUrl))),
+      });
+
+      this.client.on('error', (error) => {
+        void appInsights.trackException(error);
+      });
+    }
+  }
+
+  protected async ping(): Promise<string | undefined> {
+    return this.client?.ping();
+  }
 }
 
-export const redisCache = new RedisCache();
+export class RedisClusterCache extends RedisCacheBase {
+  readonly client?: RedisClusterType;
+
+  constructor(connectionUrl: URL) {
+    super();
+
+    /* eslint-disable @silverhand/fp/no-mutating-methods */
+    const hosts = [];
+    if (connectionUrl.host) {
+      hosts.push(connectionUrl.host);
+    }
+    hosts.push(...connectionUrl.searchParams.getAll('host'));
+    /* eslint-enable @silverhand/fp/no-mutating-methods */
+
+    const rootNodes = hosts.map((host) => {
+      return {
+        url: 'redis://' + host,
+      };
+    });
+
+    this.client = createCluster({
+      rootNodes,
+      useReplicas: true,
+      defaults: {
+        socket: this.getSocketOptions(connectionUrl),
+        username: connectionUrl.username,
+        password: connectionUrl.password,
+      },
+    });
+
+    this.client.on('error', (error) => {
+      void appInsights.trackException(error);
+    });
+  }
+
+  protected async ping(): Promise<string | undefined> {
+    return this.client?.sendCommand(undefined, true, ['PING']);
+  }
+}
+
+const redisCacheFactory = (): RedisCacheBase => {
+  const { redisUrl } = EnvSet.values;
+
+  if (redisUrl) {
+    const url = new URL(redisUrl);
+    if (yes(url.searchParams.get('redis_cluster'))) {
+      return new RedisClusterCache(url);
```

Comment:
this may cause trouble for non-normative redis urls. i made some update in [this commit](https://github.com/logto-io/logto/commit/c4c22a50fbf0a077053d537eee4cb6798eef476f). however, since you are submitting the pull from your `master` branch, i could not push to your branch directly. please feel free to cherry-pick my update. i think this is the only comment i have for this pull before merging.

<img width="569" alt="image" src="https://github.com/logto-io/logto/assets/14722250/d4b7dc56-aaa9-45b0-bcc2-2a42e0a0c8ea">

Reply:
Yeah, I realised that I submitted the PR from master branch when was too late 😅
Anyway, I cherrypicked and pushed your commit 👍

Reply:
ty